CONTRIBUTING.md

```diff
@@ -76,3 +76,115 @@ issues.
 Although not a formal channel, you can also find a number of core developers on
 the #php.pecl channel on [EFnet](http://www.efnet.org/). Similarly, many
 documentation writers can be found on #php.doc.
+
+## PHP source code directory structure
+
+PHP source code also includes several files generated during development and
+several parts where maintenance is happening upstream in their respective
+locations.
+
+```bash
+<php-src>/
+ ├─ .git/                           # Git configuration and source directory
+ └─ TSRM/                           # Thread Safe Resource Manager
+    └─ m4/                          # https://github.com/autoconf-archive/autoconf-archive
+       └─ ax_func_which_gethostbyname_r.m4
+    └─ ...
+ └─ Zend/                           # Zend Engine
+    ├─ zend_language_scanner.c      # Generated by re2c 1.0.3
+    ├─ zend_language_scanner_defs.h # Generated by re2c 1.0.3
+    ├─ zend_ini_scanner.c           # Generated by re2c 1.0.3
+    ├─ zend_ini_scanner_defs.h      # Generated by re2c 1.0.3
+    ├─ zend_vm_execute.h            # Generated by `Zend/zend_vm_gen.php`
+    ├─ zend_vm_opcodes.h            # Generated by `Zend/zend_vm_gen.php`
+    ├─ zend_vm_opcodes.c            # Generated by `Zend/zend_vm_gen.php`
+    └─ ...
+ ├─ appveyor/                       # Appveyor CI service files
+ └─ build/                          # *nix build system files
+    ├─ ax_check_compile_flag.m4     # https://github.com/autoconf-archive/autoconf-archive
+    ├─ ax_gcc_func_attribute.m4     # https://github.com/autoconf-archive/autoconf-archive
+    ├─ libtool.m4                   # https://git.savannah.gnu.org/cgit/libtool.git
+    ├─ shtool                       # https://www.gnu.org/software/shtool/
+    └─ ...
+ └─ ext/                            # PHP core extensions
+    └─ bcmath/
+       ├─ libbcmath/                # Forked and maintained in php-src
+       └─ ...
+    └─ date/
+       └─ lib/                      # Bundled datetime library https://github.com/derickr/timelib
+          ├─ parse_date.c           # Generated by re2c 0.15.3
+          ├─ parse_iso_intervals.c  # Generated by re2c 0.15.3
+          └─ ...
+       └─ ...
+    └─ fileinfo/
+       ├─ libmagic/                 # Bundled libmagic https://github.com/file/file/
+       ├─ data_file.c               # Generated by `ext/fileinfo/create_data_file.php`
+       └─ ...
+    └─ gd/
+       ├─ libgd/                    # https://github.com/libgd/libgd
```

Comment:
Looks like we don't have to submit such changes to upstream: https://github.com/php/php-src/commits/master/ext/gd/libgd

ping @cmb69

Reply:
Current `ext/gd/libgd` is a modified version of libgd 2.0.35, so it's forked and modified yet still many changes are applied upstream. Since the upstream libgd is updated frequently and is moving forward a lot it might be good to sync it according to the upstream (latest version is 2.2.5). PHP can be also built to use the system libgd (for example 2.2.5) so that's another reason to be in sync with the upstream as much as possible.

I've added as bit more explanation there `Bundled and modified GD library`...

Reply:
@carusogabriel since you've mentioned it somewhere, the `ext/curl/sync-constants.php` script is informative one and outputs the constants that needs syncing with the library and php-src... Comments and usage in the script itself is better than adding it here, I think.

Otherwise, we can mention it also if required.

Reply:
@petk Yeah, I mentioned but didn't think would be useful. But, feel free to mention it here and others scripts that might help newcomers.

Reply:
I've added also description of the `sync-constants.php` script anyway.

```bash
     └─ curl/
       ├─ sync-constants.php        # The curl symbols checker
       └─ ...
```